File: python_version/heap_sort.py
#!/usr/bin/env python3

def Heapify(arr, size, root):
    largest = root
    left = root * 2 + 1
    right = root * 2 + 2

    if left < size and arr[left] > arr[largest]:
        largest = left
    if right < size and arr[right] > arr[largest]:
        largest = right
    if largest != root:
        arr[root], arr[largest] = arr[largest], arr[root]
        Heapify(arr, size, largest)


def heap_sort(arr):
    '''sort a array'''
    size = len(arr)
    for i in range(size//2 - 1, -1, -1):
        Heapify(arr, size, i)

    for i in range(size - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        Heapify(arr, i, 0)


arr = [2, 4, 1, 5, 12, 0, 23]
print(arr)
heap_sort(arr)
print(arr)

File: cpp_version/heap_sort.cpp

#include "struct_my.h"
//Worst Case Complexity in time: O(n*log n)
//Average Case Complexity in time: O(n*log n)
//Best Case Complexity in time: O(n*log n)
//Space Complexity: O(1)
//Heap Sort Applications
//Systems concerned with security and embedded systems such as Linux Kernel use Heap Sort because of the O(n log n) upper bound on Heapsort's running time and constant O(1) upper bound on its auxiliary storage.
//Although Heap Sort has O(n log n) time complexity even for the worst case, it doesn't have more applications ( compared to other sorting algorithms like Quick Sort, Merge Sort ). However, its underlying data structure, heap, can be efficiently used if we want to extract the smallest (or largest) from the list of items without the overhead of keeping the remaining items in the sorted order. For e.g Priority Queues.
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

static Count res;

static void Heapif(vector<int> &vec, int len, int root)
{
    int largest = root;
    //左节点
    int left = 2 * root + 1;
    //右节点
    int right = 2 * root + 2;

    if (left < len && vec[left] > vec[largest])
    {
        largest = left;
    }
    if (right < len && vec[right] > vec[largest])
    {
        largest = right;
    }
    res.compare += 2;

    //使最大值在跟上
    if (largest != root)
    {
        swap(vec[largest], vec[root]);
        res.move += 3;
        //使叶子节点也满足大顶堆
        Heapif(vec, len, largest);
    }
}

Count HeapSort(vector<int> &vec)
{
    int len = vec.size();
    //最后一个有分支节点(非叶子)
    //构造大顶堆
    for (int i = len / 2 - 1; i >= 0; i--)
    {
        Heapif(vec, len, i);
    }

    //交换大顶堆和最后一个元素,使最后一个元素最大,重新heapify(但大小要减少)
    for (int i = len - 1; i > 0; i--)
    {
        res.move += 3;
        swap(vec[0], vec[i]);
        Heapif(vec, i, 0);
    }

    return res;
}
